internal/arena/unsafe.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arena
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// pointerIndex returns an integer n such that p == &s[n], or -1 if there is
+// no such integer.
+func pointerIndex[T any](p *T, s []T) int {
+	a := unsafe.Pointer(p)
+	b := unsafe.Pointer(unsafe.SliceData(s))
+	// KeepAlive escapes its argument, so this ensures that a and b have
+	// escaped to the heap and won't be moved.
+	runtime.KeepAlive([2]unsafe.Pointer{a, b})
+
+	diff := uintptr(a) - uintptr(b)
+	size := unsafe.Sizeof(*p)
+	byteLen := uintptr(len(s)) * size
+
+	// This comparison checks for the following things:
+	//
+	// 1. Obviously, that diff is not past the end of s.
+	//
+	// 2. That the subtraction did not overflow. If it did, diff will be
+	//    negative two's complement, i.e. the MSB is set, so it will be
+	//	  greater than byteLen, which, due to allocation limitations on
+	//    every platform ever, cannot be greater than MaxInt, which all
+	//	  "negative" uintptrs are greater than.
+	//
+	// 3. That byteLen is not zero. If it is zero, this branch is taken
+	//    regardless of the value of diff.
+	//
+	// 4. That p is not nil. If it is nil, then either diff will be huge
+	//    (because s is a nonempty slice) or byteLen will be zero in which case
+	//    (3) applies.
+	//
+	// Doing this as one branch is much faster than checking all four
+	// separately; this is a fairly involved strength reduction that not even
+	// LLVM can figure out in many cases, nor can Go tip as of 2024-10-28.
+	if diff >= byteLen {
+		return -1
+	}
+
+	return int(diff / size)
```

Comment:
Is there any risk that `size` might not include trailing padding bytes? In other words, is it possible that `diff/size` could return the wrong index and it should instead be something like `diff/(size+padding)`?

Reply:
Sizeof is the stride size, so it has to contain padding. https://go.dev/play/p/vYH_8QPvMTW

Consider `T = [2]byte`. If `p - s[0]` lies outside of the range of `s`, we exit early. Otherwise we perform this division, so `p` is between `s[0]` and `s[len(s)-1]`, inclusive (it is not one past, we test for that). For diff to be divisible by size, p must point to some `s[n]`, and not, say, to `&s[n] + 1`. It cannot point to `&s[n] + 1` because that would be a `*[2]byte` that does not point to an allocated `[2]byte` value (instead, it stradles two of them), which is not something Go permits. Thus, we don't need the % check.